### PR TITLE
Fix rstbx/indexing_api/tools for Python3

### LINKED ISSUE #359: rstbx/indexing_api/tools broken on Python3
to reproduce:
```
$ python -V
Python 3.6.7
$ python -c "import rstbx.indexing_api.tools"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "cctbx_project/rstbx/indexing_api/tools.py", line 40, in <module>
    spiral_order = generate_spiral_order()
  File "cctbx_project/rstbx/indexing_api/tools.py", line 36, in generate_spiral_order
    points.sort(radcmp)
TypeError: must use keyword argument for key function
```

### PR DESCRIPTION
**Pull request by Anthchirp**
fixes 39 DIALS Python 3 tests as part of dials/dials#843
resolves #359

### CODE REVIEW

**Comment by phyy-nx** [comment]
Hi, thanks for looking at this.  This pull request is a rewrite of rstbx/indexing_api/tools.py, a core component of indexing.  In my opinion, it's too much in one commit.  Here are the primary concerns:

1) It deletes functions and renames functions, so that any code that imports those functions from rstbx/indexing_api/tools.py will fail.
2) It is a large amount of altered code so it's unclear if things still work the way they used to.
3) From the commit and changes, I don't know what the python 3 error actually is so I don't know how it was fixed.

#359 states the python 3 issue can be reproduced with
`python -c "import rstbx.indexing_api.tools"`
The problem is that sort no longer uses cmp functions.  The issue can be fixed by using functools:
```
if six.PY3:
  import functools
  points.sort(key=functools.cmp_to_key(radcmp))
else:
  points.sort(radcmp) 
```

This fix is more of a kludge but it's simple and clear and it works.  A better fix would be to re-write the code to use a common comparator between py2 and py3.  I think that's what you did. 
 So I'm not saying that the changes here aren't useful or even correct, they are just too much in one commit.  If it were many commits, each changing a few things with a lot of explanation, with API changes called out and due diligence in searching repositories that could import this module, then I'd probably be able to accept that pull request.

Again, thanks for looking at this.

**Comment by Anthchirp** [comment]
Hi Aaron. Entirely reasonable. I shall break it down once I can find the time for it - may not happen before end of next week.

**Comment by Anthchirp** [comment]
This is now the same resultant file but with all changes in gory detail and with all the optimizations explained (some commits may contain a mathematical proof)

Since the essence of the file is to produce a fixed data structure I also checked the outcome by ensuring the data structure is identical to the original.

File: rstbx/indexing_api/tools.py [hunk 1]
```diff
@@ -112,36 +46,43 @@ def generate_reindex_transformations():
...
+      if sum(c*c for c in vector) > 6: break
+      if any(_is_collinear(vector, item) for item in representatives): continue
+      representatives.append(vector)
+
     reindex = []
-    for vec in vecrep:
-      for mod in modularities:
+    for vec in representatives:
+      for mod in _modularities:
+        candidate_points = [pt for pt in spiral_order if sum(v*p for v,p in zip(vec,pt))%mod == 0]
```
[the remark this answers is not in the document]
Comment (Anthchirp) [reply rstbx/indexing_api/tools.py]: No, you are absolutely correct. Good spot. Patch incoming.

**Comment by Anthchirp** [comment]
Timings list -> generator:

Python | Branch | Time
--- | --- | ---
2 | master | 17.7 msec
2 | py3rst before | 106 msec
2 | py3rst after | 11.5 msec
3 | master | **DNF**
3 | py3rst before | 109 msec
3 | py3rst after | 11.5 msec

This also fixes the issue of running out of points.

File: rstbx/indexing_api/tools.py [hunk 2]
```diff
@@ -112,36 +46,43 @@ def generate_reindex_transformations():
...
-            not is_coplanar(first,second,pt):
-            third = pt
+        for third in candidate_points:
+          if not _is_coplanar(first,second,third):
             break
-        from scitbx import matrix
-        A = matrix.sqr(first+second+third)
-        if A.determinant()<0: A = matrix.sqr(second + first + third)
+        A = scitbx.matrix.sqr(first+second+third)
+        if A.determinant()<0: A = scitbx.matrix.sqr(second + first + third)
```
[the remark this answers is not in the document]
Comment (Anthchirp) [reply rstbx/indexing_api/tools.py]: see below

File: rstbx/indexing_api/tools.py [hunk 3]
```diff
@@ -1,92 +1,26 @@
...
+import scitbx.matrix
+from rstbx.dps_core.cell_assessment import unit_cell_too_small
+from rstbx.indexing_api import cpp_absence_test
 
 # modularities 2,3,5 were sufficient for every two-image case
 # need up to 11 for Fig 4 in the single-image indexing
-modularities = [2,3,5]
-max_spiral = max(modularities)
-
-def generate_spiral_order():  #This is G0 in the paper
```
[the remark this answers is not in the document]
Comment (Anthchirp) [reply rstbx/indexing_api/tools.py]: Happy to put this back in, _but_...

Can you tell me _which_ paper this refers to?
As it stands the comment is just noise - which is why I removed it.

**Comment by Anthchirp** [comment]
For reference: I used these commands to get the timings
```bash
$ dials.python -m timeit -s "import rstbx.indexing_api.tools as T" "T.generate_spiral_order();T.generate_reindex_transformations()" # master branch
$ dials.python -m timeit -s "import rstbx.indexing_api.tools as T" "T._generate_reindex_transformations()" # py3rst branch
```

File: rstbx/indexing_api/tools.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (phyy-nx) [reply rstbx/indexing_api/tools.py]: Sauter 2004, eqn 9.

http://journals.iucr.org/j/issues/2004/03/00/dd5008/index.html#SEC2.5

Full reference: Sauter, N. K., Grosse-Kunstleve, R. W. & Adams, P. D. (2004). J. Appl. Cryst. 37, 399-409.

Adding the reference and restoring the comments would be fine.  Thanks.

[the remark this answers is not in the document]
Comment (Anthchirp) [reply rstbx/indexing_api/tools.py]: done

**Comment by phyy-nx** [comment]
Good to merge, thanks.  It would be cool if this commit history was kept, but as you like is fine :)

**Comment by Anthchirp** [comment]
24 commits may be a bit excessive for a single file :)
But I'll add a link to the pull request page for posterity

**Pull request merged**